### ISSUE TITLE
Data is accessed despite using "available_for_inference": False

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#372]
In my current setup, I am using a window function to convert the `data` input into a `padded_data_target`, which is used by the 'output' as target:

```python 
  'windowed_data': {'class': 'window', 'from': ['data'], 'stride': 3, 'window_right': 2, 'window_size': 3},
  'padded_data_target': { 'axes': ['T', 'static:0'],
                          'class': 'merge_dims',
                          'from': ['windowed_data'],
                          'n_out': 80,
                          'register_as_extern_data': 'padded_data_target'},
  'output': { 'class': 'combine',
              'from': ['dec_output', 'post_conv_tf'],
              'kind': 'add',
              'loss': 'mean_l1',
              'loss_scale': 0.25,
              'n_out': 80,
              'target': 'padded_data_target',
               },
```

`data` and `padded_data_target` are correctly labeled with `available_for_inference=False` according to the log:

```python
data: Data(shape=(None, 80), available_for_inference=False, batch_shape_meta=[...]
padded_data_target: Data(shape=(None, 80), available_for_inference=False, batch_shape_meta=[...]
```

Nevertheless, RETURNN adds `data` to the used data keys (somewhere), although is not available.
(This fails at some point in `check_matched_dataset` with some unrelated error.)

When removing the `register_as_extern_data` parameter, everything runs fine.

**Comment by albertz** [comment]
How do you define `extern_data` in your config? (But that seems fine here, `data` in `extern_data` seems correct.)

Why is `data` defined in your dataset, and why does it have dimension 0? (`<MetaDataset 'eval' epoch=1> num_outputs {'data': [0, 2], ...`) (This seems wrong.)

This error looks different to what you described to me. This seems to be some problem in `check_matched_dataset` and your dataset but not (directly) about `register_as_extern_data`. Maybe we should just ignore data keys with `available_for_inference=False` at inference in `check_matched_dataset`.

I assumed that you have an error during the construction of the layers for `register_as_extern_data`.

**Comment by JackTemaki** [comment]
No, I said that it tries to access the data which does not exist (This is why the dimension is 0), which only happens when using `register_as_extern_data`, I never said that there is anything with the layer construction.

The question is why `data` appears in the `network.used_data_keys` variable when having a layer with registered output based on `data` despite being marked as `'available_for_inference': False`
```python
extern_data = {'classes': {'dim': 77, 'shape': (None,), 'sparse': True},
               'speaker_embedding': {'dim': 128, 'shape': (None, 128)},
               'data': {'dim': 80, 'shape': (None, 80), 'available_for_inference': False}
               }
```

**Comment by albertz** [comment]
> it tries to access the data which does not exist

But I see from the output that data exists, but it has dimension 0. This seems to be the problem here. It should not exist. Dimension 0 also doesn't make sense.

> which only happens when using `register_as_extern_data`

From the error, I don't see (so far) how it is related to that. (I believe you that this might trigger it, but from the error, it looks unrelated, and probably is just an indirect effect.)

> The question is why `data` appears in the `network.used_data_keys` variable 

Why do you think this is the question here? This is again a different aspect. I don't see how this is related. This would be another separate bug.

**Comment by JackTemaki** [comment]
OggZipDataset needs to provide `data`, because otherwise `MetaDataset` does not work when a dataset does not provide `data`. This can also be considered a bug, but this is a completely different issue.

I think the only thing that has to be changed is that `register_as_extern_data` should not not trigger the creation of the layers, when the layer output is flagged as not available.

**Comment by JackTemaki** [comment]
So I found the code parts that are involved (in my opinion) in `TFNetwork`:

```python
def construct_from_dict(self, net_dict):
    """
    :param dict[str,dict[str]] net_dict:
    """
    self.layers_desc.update(net_dict)
    for name, layer_desc in sorted(net_dict.items()):
      assert isinstance(name, str)
      if name.startswith("#"):  # ignore this
        continue
      assert isinstance(layer_desc, dict)
      if layer_desc.get("register_as_extern_data"):
        # here
        self.construct_layer(net_dict, name)
    for name, layer_desc in sorted(net_dict.items()):
      assert isinstance(name, str)
      if name.startswith("#"):  # ignore this
        continue
      assert isinstance(layer_desc, dict)
      if layer_desc.get("only_on_search") and not self.search_flag:
        continue
      if layer_desc.get("only_on_eval") and not self.eval_flag:
        continue
      if (name == "output" or name.endswith(":output")
              or layer_desc.get("loss", None)
              or layer_desc.get("is_output_layer", False)):
        self.construct_layer(net_dict, name)
    assert not self._construction_stack.layers
```
The construction of the network part creating my additional target which depends on `data` is only called when `register_as_extern_data` is set. And only if this is happens, the key is added to `used_data_keys`:

```python
  def get_extern_data(self, key, mark_data_key_as_used=True):
    """
    Returns Data and add the key to self.used_data_keys if mark_data_key_as_used.
    :param str key: e.g. "data" or "classes"
    :param bool mark_data_key_as_used:
    :rtype: Data
    """
    if key in {"seq_idx", "seq_tag"} and self.parent_net:
      return self.parent_net.get_extern_data(key, mark_data_key_as_used=mark_data_key_as_used)
    if mark_data_key_as_used:
      # if key in ['seq_tag'] or self.train_flag or self.extern_data.get_data(key).available_for_inference:
      self.used_data_keys.add(key)
    if key == "seq_idx" and key not in self.extern_data.data:
      self.extern_data.data[key] = Data(
        name="seq_idx", shape=(), dtype="int32", sparse=False, auto_create_placeholders=True)
    if key == "seq_tag" and key not in self.extern_data.data:
      self.extern_data.data[key] = Data(
        name="seq_tag", shape=(), dtype="string", auto_create_placeholders=True)
    return self.extern_data.get_data(key)
```

When I prohibit that `data` is added to `used_data_keys` by having the additional check as in my commented if-line, my decoding works fine. I know that adding this check there is not possible, as it will break setups not explictely setting `available_for_inference` in the extern_data parameter, but at least this gives an indication why the problem appears.

**Comment by albertz** [comment]
> OggZipDataset needs to provide data, because otherwise MetaDataset does not work when a dataset does not provide data. This can also be considered a bug, but this is a completely different issue.

This sounds like the main bug/problem here. Why do you think this is a different issue? The issue here (the exception you posted) is about incorrect data in the dataset. Or maybe you posted a wrong/irrelevant exception?

All what you say about `register_as_extern_data` sounds like another problem, not about this exception, which is about incorrect `data` in the dataset.

**Comment by JackTemaki** [comment]
> > OggZipDataset needs to provide data, because otherwise MetaDataset does not work when a dataset does not provide data. This can also be considered a bug, but this is a completely different issue.
> 
> This sounds like the main bug/problem here. Why do you think this is a different issue? The issue here (the exception you posted) is about incorrect data in the dataset. Or maybe you posted a wrong/irrelevant exception? The purpose of `check_matched_dataset` is to check whether your definition of `extern_data` matches what the dataset gives you, and that fails here because the dataset defines `data` incorrectly (it should not define it at all, as it does not provide it).

If `data` would not exists at all, `check_matched_dataset` would still fail, as it requires that the dataset provides all keys that are in `used_data_keys`. This is why the exception message itself is not relevant here, but the reason for the exception. `data` should NOT be in `TFNetwork.used_data_keys`. This is incorrect behaviour, and the main problem here.

> All what you say about `register_as_extern_data` sounds like another problem, not about this exception, which is about incorrect `data` in the dataset.

Would the issue have been more clear without any exception? I am not sure about that. I said the incorrect access to `data` is the problem, not the resulting rxception because of the incorrect access.

**Comment by albertz** [comment]
> This is why the exception message itself is not relevant here

Ok, I think then the exception here is misleading. I thought this issue is about the exception.

> Would the issue have been more clear without any exception?

Yes, definitely, if the exception is about some unrelated problem (namely that `data` has dimension 0). I thought this is the problem you want to highlight here, that there is a mismatch of `data` in the dataset vs `data` in `extern_data`, i.e. exactly what the error says.

I will edit it.

**Comment by albertz** [comment]
Just to clarify, so far you have not debugged where/why exactly RETURNN adds `data` to the used data keys? Ie. the stack trace when it calls `get_extern_data`?

**Comment by JackTemaki** [comment]
I did not paste the full stack trace, but I posted the two relevant code parts that are part of the stack trace. 

From the code parts above, this was the "starting point" of where the working models differs from the not-working model.
```python
if layer_desc.get("register_as_extern_data"):
        # here
        self.construct_layer(net_dict, name)
```
So using "register_as_extern_data" leads to the call of `self.used_data_keys.add(key)`.

Let me collect the stack trace and put it in a GIST.

**Comment by albertz** [comment]
I mean, these code parts are clear (and not so interesting), because (I think) that's the only place where it could have added to `used_data_keys`. The more relevant and interesting question is, where does this function (`get_extern_data`) gets called, and what is the stack trace. This seems to be wrong. It should not be called for `data`, or only with `mark_data_key_as_used=False`. Because `data` should not be used. The question is, where is it used.

But you already answered it with `register_as_extern_data`, at least the starting point (and I assume that data is used somewhere by these layers). This is also what I already mentioned in my first comment. During the construction, we should skip layers which have `layer.output.available_for_inference` = False, if this is not evaluation (eval flag False).

**Comment by JackTemaki** [comment]
> I mean, these code parts are clear (and not so interesting), because (I think) that's the only place where it could have added to `used_data_keys`. The more relevant and interesting question is, where does this function (`get_extern_data`) gets called, and what is the stack trace. This seems to be wrong. It should not be called for `data`, or only with `mark_data_key_as_used=False`. Because `data` should not be used. The question is, where is it used.

This is in the Source Layer. I am not sure if we cause side effects when we check there.

> But you already answered it with `register_as_extern_data`, at least the starting point (and I assume that data is used somewhere by these layers). This is also what I already mentioned in my first comment. During the construction, we should skip layers which have `layer.output.available_for_inference` = False, if this is not evaluation (eval flag False).

But do we know this at that point? During the construction it is unknown, that one of the inputs will not be available, right?

**Comment by albertz** [comment]
> This is in the Source Layer. I am not sure if we cause side effects when we check there.

What check do you mean?
But why is it used in the SourceLayer? This seems wrong. It should not use `data` in your setup.

> But do we know [eval flag] at that point? During the construction it is unknown, that one of the inputs will not be available, right?

Sure we know this. During construction, the flags (train, eval, search) are already given.

Also, we know the `available_for_inference` flag during construction. This is defined by `extern_data`.

> Also, all extern_data entries besides data get available_for_inference=False as default, so editing anything here based on the availability will result in breaking all old configs.

I don't understand what you mean. You define this in your config.

**Comment by JackTemaki** [comment]
> > This is in the Source Layer. I am not sure if we cause side effects when we check there.
> 
> What check do you mean?
> But why is it used in the SourceLayer? This seems wrong. It should not use `data` in your setup.

But we already know "why" it used. The question is at which point we fix the code.

> 
> > But do we know [eval flag] at that point? During the construction it is unknown, that one of the inputs will not be available, right?
> 
> Sure we know this. During construction, the flags (train, eval, search) are already given.

Yes, but do we already know at that point that any of the dependencies is not "available"?

> 
> Also, we know the `available_for_inference` flag during construction. This is defined by `extern_data`.
> 
> > Also, all extern_data entries besides data get available_for_inference=False as default, so editing anything here based on the availability will result in breaking all old configs.
> 
> I don't understand what you mean. You define this in your config.

Yes, I can define it to be correct. But the default had wrong behaviour. You could say I only wrote "wrong" configs in the last two years. But they all worked. If I fix this now, none of them will work. For me this would be fine as I have fixed RETURNN commit dependencies, but this will not be fine for other people.

So if the intended behaviour should be "An extern data object is not available for inference -> the data source is never added to used keys from a source layer", this will be a problem.

At least I think it should have been that way.

**Comment by albertz** [comment]
> But we already know "why" it used. The question is at which point we fix the code.

I got confused about your mention of SourceLayer. There is no problem of course in SourceLayer (and also no way how you would fix it at that point). SourceLayer is behaving totally correct.

Anyway, all the layers are all irrelevant and have the same problem in the same way, whenever they access some data which is not available. And there is nothing wrong in the layers.

The question is also not too much at which point this should be fixed. I already outlined roughly how to do it.

> Yes, but do we already know at that point that any of the dependencies is not "available"?

That is what I explained, what we can derive. It's actually simple:
During the network construction, we can skip layers which have `layer.output.available_for_inference` = False, if this is not evaluation (eval flag False).
Or sth similar. In any case, it's very clear that this has to be fixed at this point.
I can do that later. I think this is very simple. Maybe you can prepare a small test case.

>>> Also, all extern_data entries besides data get available_for_inference=False as default, so editing anything here
>>> based on the availability will result in breaking all old configs.
>>
>> I don't understand what you mean. You define this in your config.
>
> Yes, I can define it to be correct. But the default had wrong behaviour.

I don't understand. Why wrong? Any default setting is sometimes what you want, and sometimes not what you want. But why is the default behavior wrong? A default behavior can never be wrong. It can just maybe be suboptimal, or unintuitive and not match the common case. But I don't think this is even the case here, right? The default is that "data" is available at inference, "classes" not. I think this is a pretty reasonable default.

> You could say I only wrote "wrong"
> configs in the last two years. But they all worked. If I fix this now, none of them will work.

What do you mean? What do you want to fix? What exactly do you think it would break?

Using any non-available data has never worked. Why would you break sth?

> So if the intended behaviour should be "An extern data object is not available for inference -> the data source is never added to used keys from a source layer", this will be a problem.

I don't understand. This is not at all about the source layer. Why is the source layer relevant?

Also, I don't understand what you mean. If you try to use data which is not available, this obviously will result in an error. This was always the case. It cannot possibly work, because the data is not available.

Or do you mean you have the data actually available at inference but incorrectly defined `available_for_inference = False` in your config? But why would you even do that in the first place? If this is the case, I think you have to fix your configs, as they are obviously wrong. We cannot and should not try to keep compatible to obvious bugs.

Or do you mean the data is available, but the current default is wrong, and you have not specified `available_for_inference` in your config? This is yet another case, also somewhat a separate issue (we should not mix it in this issue here, i.e. I think better open a separate issue about this). Also, I think this is not too much a problem. I think we can come up with some solutions. E.g. some special handling for `available_for_inference=NotSpecified`, or maybe derive it from the dataset if not specified, or sth like that.

**Comment by JackTemaki** [comment]
> > Yes, but do we already know at that point that any of the dependencies is not "available"?
> 
> That is what I explained, what we can derive. It's actually simple:
> During the network construction, we can skip layers which have `layer.output.available_for_inference` = False, if this is not evaluation (eval flag False).

Ok

> Or do you mean you have the data actually available at inference but incorrectly defined `available_for_inference = False` in your config? But why would you even do that in the first place? If this is the case, I think you have to fix your configs, as they are obviously wrong. We cannot and should not try to keep compatible to obvious bugs.
> Or do you mean the data is available, but the current default is wrong, and you have not specified `available_for_inference` in your config? This is yet another case, also somewhat a separate issue (we should not mix it in this issue here, i.e. I think better open a separate issue about this). Also, I think this is not too much a problem. I think we can come up with some solutions. E.g. some special handling for `available_for_inference=NotSpecified`, or maybe derive it from the dataset if not specified, or sth like that.

This is exactly what I mean with the wrong defaults. As the default behaviour was absolutely not intuitive (why would any data I explicitly define not be available per default?). I just say we can not fix this issue without also fixing the default case. I do not understand why this should not be "mixed" in with this issue?

**Comment by albertz** [comment]
> I do not understand why this should not be "mixed" in with this issue?

This issue here is about wrong handling for `available_for_inference = False`. The fix for this would be the extended logic in the network construction, as explained above.

The issue about default setting of `available_for_inference` in `extern_data`, if not specified by the user, is separate, and also needs a separate fix, like some new logic of the default behavior. It's independent of the wrong handling of `available_for_inference = False`.

Both issues might be related, but they are really separate, and need separate fixes. Even the discussion about how to fix them or deal with them is independent/separate of each other. We just make the discussion about the problem itself, and how to fix it, more complicated by mixing all the discussions.

If we fix the `available_for_inference = False` handling, we have not fixed the default behavior.
If we fix the default behavior, we have not fixed the `available_for_inference = False` handling.

An issue should ideally be about a single thing, a single thing which you can fix. An issue should not need to require multiple fixes, because that always means that there are actually multiple issues. At least then there should be separate issues about each individual issue, to easier discuss about each individual issue, and how to proceed/solve it.

See also how other bigger projects do this. If they have one issue covering multiple smaller issues, then they always have also the smaller issues as separate issues, and the bigger/combined issue is blocked by the small/atomic/separate issues.

**Comment by JackTemaki** [comment]
> > the default behaviour was absolutely not intuitive (why would any data I explicitly define not be available per default?)
> 
> You always need to explicitly define all the data. Originally it was `num_inputs`/`num_outputs` (for `data`/`classes`) and then it was generalized by `extern_data`. `num_outputs` already tells by the name that this is the output. I think it's really the standard case that the output (the targets) are not available during inference.

Yes, this might be historically true, I am only saying if the variable is called "extern_data", it does not look like any data defined here other than "data" will not be available during inference.

> 
> > I do not understand why this should not be "mixed" in with this issue?
> 
> This issue here is about wrong handling for `available_for_inference = False`. The fix for this would be the extended logic in the network construction, as explained above.
> 
> The issue about default setting of `available_for_inference` in `extern_data`, if not specified by the user, is separate, and also needs a separate fix, like some new logic of the default behavior. It's independent of the wrong handling of `available_for_inference = False`.
> 
> Both issues might be related, but they are really separate, and need separate fixes. Even the discussion about how to fix them or deal with them is independent/separate of each other. We just make the discussion about the problem itself, and how to fix it, more complicated by mixing all the discussions.
> 
> If we fix the `available_for_inference = False` handling, we have not fixed the default behavior.
> If we fix the default behavior, we have not fixed the `available_for_inference = False` handling.
> 
> An issue should ideally be about a single thing, a single thing which you can fix. An issue should not need to require multiple fixes, because that always means that there are actually multiple issues. At least then there should be separate issues about each individual issue, to easier discuss about each individual issue, and how to proceed/solve it.
> 
> See also how other bigger projects do this. If they have one issue covering multiple smaller issues, then they always have also the smaller issues as separate issues, and the bigger/combined issue is blocked by the small/atomic/separate issues.

Well, ok then.

Then to this issue. I did not find any position in the code, where I have access to the `layer` object during construction, but also the general call order is not clear to me. It would be nice if the docstrings of the functions of the `TFNetwork` class would get more meaningful docstrings.

So from what I see, I can check for this in `TFNetwork._create_layer`, but here I am not allowed to return "None".

I think there should be some way to access the result of the template construction (which would contain the information about `layer.output.available_for_inference`) but I just do not see it.

**Comment by albertz** [comment]
> Then to this issue. I did not find any position in the code, where I have access to the `layer` object during construction,

You have access to the `output` template (`output_template` in the code), which is what I meant. The layer itself of course must not (can not) be created, as the data is not available.

> but also the general call order is not clear to me. It would be nice if the docstrings of the functions of the `TFNetwork` class would get more meaningful docstrings.

What exactly is unclear?
 
> So from what I see, I can check for this in `TFNetwork._create_layer`, but here I am not allowed to return "None".

You could just create and return a dummy layer, with this template output. We have `InternalLayer` for that. Or maybe we could introduce this class:
```
class UnavailableDataLayer(InternalLayer):
  """ .... """
```

> I think there should be some way to access the result of the template construction (which would contain the information about `layer.output.available_for_inference`) but I just do not see it.

But there is. It's `output_template`, just in the function you mentioned, in `_create_layer`.

**Comment by JackTemaki** [comment]
> > Then to this issue. I did not find any position in the code, where I have access to the `layer` object during construction,
> 
> You have access to the `output` template (`output_template` in the code), which is what I meant. The layer itself of course must not (can not) be created, as the data is not available.
> 
> > but also the general call order is not clear to me. It would be nice if the docstrings of the functions of the `TFNetwork` class would get more meaningful docstrings.
> 
> What exactly is unclear?

If I have time I will add docstrings in a PR, then you can comment and correct them.

> 
> > So from what I see, I can check for this in `TFNetwork._create_layer`, but here I am not allowed to return "None".
> 
> You could just create and return a dummy layer, with this template output. We have `InternalLayer` for that. Or maybe we could introduce this class:
> 
> ```
> class UnavailableDataLayer(InternalLayer):
>   """ .... """
> ```
> 
> > I think there should be some way to access the result of the template construction (which would contain the information about `layer.output.available_for_inference`) but I just do not see it.
> 
> But there is. It's `output_template`, just in the function you mentioned, in `_create_layer`.

Yes, but this is why I asked for what I can return there. I did not know returning a "dummy" layer is a valid possibility.

I will check this approach.

**Comment by albertz** [comment]
> I did not know returning a "dummy" layer is a valid possibility.

Why not? Who says this should not be valid? The `output` of this layer is anyway not available.

It might be a problem for subnetworks, though. Although we can maybe extend it later. But this might become a bit hacky/ugly.

Or, the other possibility would be, to not construct the layer at all. We could throw a `DependsOnUnavailableData` exception whenever we get this. This could even be thrown directly in `get_extern_data` with `mark_data_key_as_used=True`. But then you would not have the output template available for `register_as_extern_data`, so this is not really an option.

We could maybe also use `_TemplateLayer`. And even maybe reuse parts/code of the template layer construction. E.g. see `SubnetworkLayer._construct_template_subnet`.
I think this is buggy for further subnets as well, though, in a very similar way.

**Comment by JackTemaki** [comment]
fixed with #374
